src/infra/dataset-update-flow-inmem/src/dataset_flow_key.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Borrow;
+use std::hash::{Hash, Hasher};
+
+use kamu_dataset_update_flow::DatasetFlowType;
+use opendatafabric::DatasetID;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Eq, PartialEq, Hash, Clone)]
+pub(crate) struct OwnedDatasetFlowKey {
+    dataset_id: DatasetID,
+    flow_type: DatasetFlowType,
+}
+
+impl OwnedDatasetFlowKey {
+    pub fn new(dataset_id: DatasetID, flow_type: DatasetFlowType) -> Self {
+        Self {
+            dataset_id,
+            flow_type,
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Copy, Clone, Hash, PartialEq, Eq)]
+pub(crate) struct BorrowedDatasetFlowKey<'a> {
+    dataset_id: &'a DatasetID,
```

Comment:
I did not know it's a type that has no heap pointers, and, if so, why doesn't it implement `Copy` trait? It would simplify half of the codebase.

Still, it's a 48-byte structure, and copying it around seems to be thick in theory. Copying 12 bytes seems to be faster, but that is not guaranteed as well, as memory is read by pages anyway. At least, it won't fit into registers.

I would say "cheap copy" if it had `Rc<>` over a 32-bit buffer inside, but in our case such approach will require `Arc`, and that is unlikely cheaper than copying 48 bytes.

Reply:
I'm going to keep this little optimization anyway. We need to have a working example of multi-field hashing :-)